### PR TITLE
[A2-751] Allow MSAD config to override all defaults

### PR DESCRIPTION
**Pull request by tylercloke**
### :nut_and_bolt: Description

MSAD config has some defaults values set up that allow it to be easier to configure for MSAD specific servers. However, you could not override some of those defaults, which meant that if there was a default that didn't work for your specific MSAD setup, you'd have to switch over to the LDAP config. In doing so, you'd have to set up all the MSAD defaults in LDAP, even though you just wanted to override a single default.

Now, we allow the user to override any default they wish in the MSAD config, while maintaining the sane MSAD defaults if the value is not specified.

### :athletic_shoe: Demo Script / Repro Steps

```
rebuild components/automate-dex && rebuild components/automate-gateway/ && rebuild components/automate-deployment/ && rebuild components/automate-cli
chef-automate config show > test.toml
```

Insert bare minimum into test.toml MSAD portion:

```
[dex]
  [dex.v1]
    [dex.v1.sys]
      [dex.v1.sys.connectors]
      [dex.v1.sys.connectors.msad_ldap]
host = "172.21.121.71:389"
bind_dn = "CN=Admin,OU=Users,OU=authteam,DC=authteam,DC=example,DC=com"
bind_password = "ChefAutomate1"
base_user_search_dn = "DC=authteam,DC=example,DC=com"
base_group_search_dn = "DC=authteam,DC=example,DC=com"
```

Resulting config still contains sane MSAD defaults (the same as master):

```
connectors:
  - type: ldap
    id: ldap
    name: LDAP Microsoft AD
    config:
      host: 172.21.121.71:389
      bindDN: $LDAP_BINDDN
      bindPW: $LDAP_BINDPW
      insecureSkipVerify: true
      userSearch:
        baseDN: DC=authteam,DC=example,DC=com
        filter: (objectClass=person)
        username: sAMAccountName
        idAttr: sAMAccountName
        emailAttr: mail
        nameAttr: displayName
      groupSearch:
        baseDN: DC=authteam,DC=example,DC=com
        filter: (objectClass=group)
        userAttr: DN
        groupAttr: member
        nameAttr: displayName
```

Override :allthethings: in test.toml

```
[dex]
  [dex.v1]
    [dex.v1.sys]
      [dex.v1.sys.connectors]
      [dex.v1.sys.connectors.msad_ldap]
host = "172.21.121.71:389"
bind_dn = "CN=Admin,OU=Users,OU=authteam,DC=authteam,DC=example,DC=com"
bind_password = "ChefAutomate1"
base_user_search_dn = "DC=authteam,DC=example,DC=com"
base_group_search_dn = "DC=authteam,DC=example,DC=com"
# overrides
username_attr = "username_attr"
user_id_attr = "user_id_attr"
insecure_no_ssl = true
user_query_filter = "user_query_filter"
email_attr = "email_attr"
user_display_name_attr = "user_display_name_attr"
group_query_filter = "group_query_filter"
filter_groups_by_user_value = "filter_groups_by_user_value"
filter_groups_by_user_attr = "filter_groups_by_user_attr"
group_display_name_attr = "group_display_name_attr"
```

Results in:

```
connectors:
  - type: ldap
    id: ldap
    name: LDAP Microsoft AD
    config:
      host: 172.21.121.71:389
      bindDN: $LDAP_BINDDN
      bindPW: $LDAP_BINDPW
      insecureSkipVerify: true
      insecureNoSSL: true
      userSearch:
        baseDN: DC=authteam,DC=example,DC=com
        filter: user_query_filter
        username: username_attr
        idAttr: user_id_attr
        emailAttr: email_attr
        nameAttr: user_display_name_attr
      groupSearch:
        baseDN: DC=authteam,DC=example,DC=com
        filter: group_query_filter
        userAttr: filter_groups_by_user_value
        groupAttr: filter_groups_by_user_attr
        nameAttr: group_display_name_attr
```

### :white_check_mark: Checklist

- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [ ] Code actually executed?
- [ ] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/automate-chef-io/content/docs/configuration.md [hunk 1]
```diff
@@ -261,32 +263,45 @@ bind_dn = "<your bind_dn>"
...
-ca_contents = "<your ca contents>"
-user_query_filter = "<your filter for user queries>"
-group_query_filter = "<your filter for group queries>"
+ca_contents = "<your ca contents>" # optional, but recommended
 ```
 
 The MSAD configuration is an LDAP configuration with more provided default
-values that are commonly a good fit for Active Directory:
+values that are commonly a good fit for Active Directory, but you can override
+any one but uncommenting it in the config and setting its value:
```
Comment (bcmdarroch) [inline components/automate-chef-io/content/docs/configuration.md]: s/but/by

File: components/automate-chef-io/content/docs/configuration.md [hunk 2]
```diff
@@ -243,16 +243,18 @@ policies will be incorrectly labelled as `ldap`.
...
 Chef Automate's default configuration for Microsoft AD is specific to LDAP.
 To configure Microsoft AD using SAML, see the SAML configuration section below.
 {{< /warning >}}
 
-In this config, all fields are required except for `ca_contents`, `user_query_filter`,
-`base_group_search_dn` and `group_query_filter`. See [LDAP]({{< relref "ldap.md" >}}) integration for more information.
-goes into further detail about what each of those config fields represents.
+In this config, the required fields are `host`, `bind_dn`,
+`bind_password`, `base_user_search_dn`, and `base_group_search_dn`.
+See [LDAP]({{< relref "ldap.md" >}}) integration for more information.
```
Comment (susanev) [inline components/automate-chef-io/content/docs/configuration.md]:
```suggestion
See [LDAP Integration]({{< relref "ldap.md" >}}) for more information.
```

File: api/config/dex/config_request.proto [hunk 3]
```diff
@@ -105,6 +105,14 @@ message ConfigRequest {
 			google.protobuf.StringValue base_group_search_dn = 6;
 			google.protobuf.StringValue user_query_filter = 7;
 			google.protobuf.StringValue group_query_filter = 8;
+			google.protobuf.StringValue username_attr = 9;
```
Comment (msorens) [inline api/config/dex/config_request.proto]: Do you have a reference that specifies all these fields? Might be nice to include a link in this file.

File: api/config/dex/config_request_test.go [hunk 4]
```diff
@@ -430,6 +432,85 @@ func TestValidate(t *testing.T) {
...
+
+		// User defined
+		assert.Equal(t, "ad.au", c.GetConnectors().GetMsadLdap().GetHost().GetValue())
+		assert.Equal(t, "cn=admin,dc=org", c.GetConnectors().GetMsadLdap().GetBindDn().GetValue())
+		assert.Equal(t, "ChefAutomate1", c.GetConnectors().GetMsadLdap().GetBindPassword().GetValue())
+		assert.Equal(t, "ChefAutomate1", c.GetConnectors().GetMsadLdap().GetBindPassword().GetValue())
+		assert.Equal(t, "ou=people,dc=org", c.GetConnectors().GetMsadLdap().GetBaseUserSearchDn().GetValue())
+		assert.Equal(t, "ou=groups,dc=org", c.GetConnectors().GetMsadLdap().GetBaseGroupSearchDn().GetValue())
+
+		// Defaults are unset so that they recieve overrides in config
```
Comment (msorens) [inline api/config/dex/config_request_test.go]: s/recieve/receive/

File: api/config/dex/config_request.proto (hunk 3 above)
Comment (tylercloke) [reply api/config/dex/config_request.proto]: They eventually map to LDAP config settings in Dex but they aren't 1-1.

File: components/automate-chef-io/content/docs/configuration.md [hunk 5]
```diff
@@ -243,16 +243,18 @@ policies will be incorrectly labelled as `ldap`.
 
 Chef Automate comes with a default LDAP configuration for Microsoft Active Directory (MSAD).
 The MSAD configuration is intended for getting started with minimal configuration for standard MSAD systems.
-If you need to customize your configuration further, see the [LDAP section]({{< relref "configuration.md#LDAP" >}}) below.
+The MSAD configuration makes some assumptions about your MSAD setup, but you can override those defaults.
 
 {{< warning >}}
 Chef Automate's default configuration for Microsoft AD is specific to LDAP.
 To configure Microsoft AD using SAML, see the SAML configuration section below.
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]: Please include an internal link to the SAML configuration section for ease of reference.
```suggestion
To configure Microsoft AD using SAML, see the SAML configuration section.
```

File: components/automate-chef-io/content/docs/configuration.md [hunk 6]
```diff
@@ -243,16 +243,18 @@ policies will be incorrectly labelled as `ldap`.
 
 Chef Automate comes with a default LDAP configuration for Microsoft Active Directory (MSAD).
 The MSAD configuration is intended for getting started with minimal configuration for standard MSAD systems.
-If you need to customize your configuration further, see the [LDAP section]({{< relref "configuration.md#LDAP" >}}) below.
+The MSAD configuration makes some assumptions about your MSAD setup, but you can override those defaults.
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]:
```suggestion
Default values in this MSAD configuration can be overridden.
```

File: components/automate-chef-io/content/docs/configuration.md [hunk 7]
```diff
@@ -243,16 +243,18 @@ policies will be incorrectly labelled as `ldap`.
...
 Chef Automate's default configuration for Microsoft AD is specific to LDAP.
 To configure Microsoft AD using SAML, see the SAML configuration section below.
 {{< /warning >}}
 
-In this config, all fields are required except for `ca_contents`, `user_query_filter`,
-`base_group_search_dn` and `group_query_filter`. See [LDAP]({{< relref "ldap.md" >}}) integration for more information.
-goes into further detail about what each of those config fields represents.
+In this config, the required fields are `host`, `bind_dn`,
+`bind_password`, `base_user_search_dn`, and `base_group_search_dn`.
+See [LDAP Integration]({{< relref "ldap.md" >}}) for more information.
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]:
```suggestion
See [LDAP Integration]({{< relref "ldap.md" >}}) for more information on config fields, including optional fields.
```

File: components/automate-chef-io/content/docs/configuration.md [hunk 8]
```diff
@@ -243,16 +243,18 @@ policies will be incorrectly labelled as `ldap`.
...
+It goes into further detail about what each of those config fields represents,
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]:
```suggestion
```

File: components/automate-chef-io/content/docs/configuration.md [hunk 9]
```diff
@@ -243,16 +243,18 @@ policies will be incorrectly labelled as `ldap`.
...
+including optional fields.
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]:
```suggestion
```

File: components/automate-chef-io/content/docs/configuration.md [hunk 10]
```diff
@@ -261,32 +263,44 @@ bind_dn = "<your bind_dn>"
...
 base_group_search_dn = "<your base group search DN>"
-ca_contents = "<your ca contents>"
-user_query_filter = "<your filter for user queries>"
-group_query_filter = "<your filter for group queries>"
+ca_contents = "<your ca contents>" # optional, but recommended
 ```
 
 The MSAD configuration is an LDAP configuration with more provided default
-values that are commonly a good fit for Active Directory:
+values that are commonly a good fit for Active Directory, but you can override any one by uncommenting it in the config and setting its value:
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]:
```suggestion
values that are commonly a good fit for Active Directory. Override any single default value by uncommenting it in the config and setting its value:
```

File: components/automate-chef-io/content/docs/configuration.md [hunk 11]
```diff
@@ -261,32 +263,44 @@ bind_dn = "<your bind_dn>"
...
+[dex.v1.sys.connectors.msad_ldap]
+host = "<your host>"
+bind_dn = "<your bind_dn>"
+bind_password = "<your bind_password>"
+base_user_search_dn = "<your base user search DN>"
+base_group_search_dn = "<your base group search DN>"
+ca_contents = "<your ca contents>" # optional
+
+# MSAD default values (uncomment to override a specific one)
+# insecure_no_ssl = false
```
Comment (mjingle) [inline components/automate-chef-io/content/docs/configuration.md]: Do these commented-out examples need to have formatting like `"<sAMAccountName>"` with the < > brackets?

Comment (tylercloke) [reply components/automate-chef-io/content/docs/configuration.md]: Not in this case. Those are real values that are the defaults.

**Pull request merged**